File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeLists(ListNode* &head1, ListNode* &head2){
        ListNode* dummy= new ListNode(-1);
        ListNode* ptr1=head1;
        ListNode* ptr2= head2;
        ListNode* ptr3=dummy;
        while(ptr1!=NULL && ptr2!=NULL){
            if(ptr1->val < ptr2->val){
                ptr3->next=ptr1;
                ptr1= ptr1->next;
            }
            else{
                ptr3->next=ptr2;
                ptr2=ptr2->next;
            }
            ptr3=ptr3->next;
        }
        if(ptr1!=NULL) ptr3->next=ptr1;
        if(ptr2!=NULL) ptr3->next=ptr2;
        return dummy->next;
    }
    
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0) return NULL;
        while(lists.size()>1){
            ListNode* mergeHead=  mergeLists(lists[0], lists[1]);
            lists.push_back(mergeHead);
            lists.erase(lists.begin());
            lists.erase(lists.begin());
        }
        return lists[0];
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists==null || lists.length==0) return null;
        return mergeKLists(lists, 0, lists.length-1);
    }
    
    public ListNode mergeKLists(ListNode[] lists, int start, int end){
        if(start==end) return lists[start];
        
        int mid= start+(end-start)/2;
        ListNode left=mergeKLists(lists,start, mid);
        ListNode right= mergeKLists(lists, mid+1, end);
        return mergeTwoLists(left, right);
    }
    
    public ListNode mergeTwoLists(ListNode list1, ListNode list2){
        ListNode dummy= new ListNode(0);
        ListNode curr= dummy;
        while(list1!=null && list2!=null){
            if(list1.val<=list2.val){
                curr.next=list1;
                list1= list1.next;
            }
            
            else{
                curr.next= list2;
                list2= list2.next;
            }
            curr= curr.next;
        }
        if(list1!=null){
            curr.next=list1;
        }
        else{
            curr.next=list2;
        }
        return dummy.next;
    }
}